job/server/src/main/java/alluxio/master/job/workflow/WorkflowTracker.java

```diff
@@ -77,6 +77,27 @@ public synchronized void run(WorkflowConfig workflowConfig, long jobId)
     next(jobId);
   }
 
+  /**
+   * Cancels a job with a particular job id.
+   * @param jobId the job id
+   * @return true if job exists, false otherwise
+   */
+  public synchronized boolean cancel(long jobId) {
+    ConcurrentHashSet<Long> children = mChildren.get(jobId);
+    if (children == null) {
+      return false;
+    }
+
+    for (Long child : children) {
+      try {
+        mJobMaster.cancel(child);
+      } catch (JobDoesNotExistException e) {
+        LOG.info("Tried to cancel jobId: {} but the job did not exist", child);
+      }
+    }
+    return true;
+  }
```

Comment:
What are the guarantees for canceling a workflow in different cases, for example if some children have succeeded? Will we ever have partial job completion/cancellation?

Reply:
That is correct. In fact, even if this job has already failed, there may be some tasks that are running in parallel which this would also cancel.

Reply:
In the case that the job has completed some tasks and we cancel the rest, will there be partial state, or are the guarantees up to each individual job to implement?

Reply:
It cancels everything it "can". In other words, it goes to all of the leaves (tasks) and cancels those. 

Even for plan cancelling, it just cancels everything it has.